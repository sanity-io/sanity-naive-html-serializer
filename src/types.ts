import {ObjectField, SanityDocument, TypedObject, Schema} from 'sanity'

export type SerializedDocument = {
  name: string
  content: string
}

export type TranslationLevel = 'document' | 'field'
export interface Serializer {
  serializeDocument: (
    doc: SanityDocument,
    translationLevel: TranslationLevel,
    baseLang?: string,
    stopTypes?: string[],
    serializers?: Record<string, any>
  ) => SerializedDocument
  fieldFilter: (
    obj: Record<string, any>,
    objFields: ObjectField[],
    stopTypes: string[]
  ) => TypedObject
  languageObjectFieldFilter: (obj: Record<string, any>, baseLang: string) => Record<string, any>
  serializeArray: (
    fieldContent: Record<string, any>[],
    fieldName: string,
    stopTypes: string[],
    serializers: Record<string, any>
  ) => string
  serializeObject: (
    obj: TypedObject,
    stopTypes: string[],
    serializers: Record<string, any>
  ) => string
}

export type SerializerClosure = (schemas: Schema) => Serializer

export interface Deserializer {
  deserializeDocument: (
    serializedDoc: string,
    deserializers?: Record<string, any>,
    blockDeserializers?: Array<any>
  ) => Record<string, any>
  deserializeHTML: (
    html: string,
    deserializers: Record<string, any>,
    blockDeserializers: Array<any>
  ) => Record<string, any> | any[]
}

export interface Merger {
  fieldLevelMerge: (
    translatedFields: Record<string, any>,
    baseDoc: SanityDocument,
    localeId: string,
    baseLang: string
  ) => Record<string, any>
  documentLevelMerge: (
    translatedFields: Record<string, any>,
    baseDoc: SanityDocument
  ) => Record<string, any>
  reconcileArray: (origArray: any[], translatedArray: any[]) => any[]
  reconcileObject: (
    origObject: Record<string, any>,
    translatedObject: Record<string, any>
  ) => Record<string, any>
}
