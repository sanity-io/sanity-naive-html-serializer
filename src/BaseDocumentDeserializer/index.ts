import BaseDocumentDeserializer from './BaseDocumentDeserializer'
import LegacyBaseDocumentDeserializer from './LegacyBaseDocumentDeserializer'

export { LegacyBaseDocumentDeserializer, BaseDocumentDeserializer }
