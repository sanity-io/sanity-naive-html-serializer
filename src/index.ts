export {BaseDocumentMerger} from './BaseDocumentMerger'
export {BaseDocumentSerializer} from './BaseDocumentSerializer'
export {BaseDocumentDeserializer} from './BaseDocumentDeserializer'
export {
  defaultStopTypes,
  customSerializers,
  customBlockDeserializers,
} from './BaseSerializationConfig'

export type {SerializedDocument, Serializer, SerializerClosure, Deserializer, Merger} from './types'
