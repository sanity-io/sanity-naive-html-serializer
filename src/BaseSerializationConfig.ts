import {PortableTextBlockStyle} from '@portabletext/types'

import {
  PortableTextBlockComponent,
  PortableTextListComponent,
  PortableTextListItemComponent,
} from '@portabletext/to-html'

import {htmlToBlocks} from '@sanity/block-tools'
import {blockContentType} from './BaseDocumentDeserializer/helpers'
import {PortableTextTextBlock, TypedObject} from 'sanity'

export const defaultStopTypes = [
  'reference',
  'date',
  'datetime',
  'file',
  'geopoint',
  'image',
  'number',
  'crop',
  'hotspot',
  'boolean',
  'url',
  'color',
  'code',
]

export const defaultPortableTextBlockStyles: Record<
  PortableTextBlockStyle,
  PortableTextBlockComponent | undefined
> = {
  normal: ({value, children}) => `<p id="${value._key}">${children}</p>`,
  blockquote: ({value, children}) => `<blockquote id="${value._key}">${children}</blockquote>`,
  h1: ({value, children}) => `<h1 id="${value._key}">${children}</h1>`,
  h2: ({value, children}) => `<h2 id="${value._key}">${children}</h2>`,
  h3: ({value, children}) => `<h3 id="${value._key}">${children}</h3>`,
  h4: ({value, children}) => `<h4 id="${value._key}">${children}</h4>`,
  h5: ({value, children}) => `<h5 id="${value._key}">${children}</h5>`,
  h6: ({value, children}) => `<h6 id="${value._key}">${children}</h6>`,
}

const defaultLists: Record<'number' | 'bullet', PortableTextListComponent> = {
  number: ({value, children}) => `<ol id="${value._key.replace('-parent', '')}">${children}</ol>`,
  bullet: ({value, children}) => `<ul id="${value._key.replace('-parent', '')}">${children}</ul>`,
}

const defaultListItem: PortableTextListItemComponent = ({value, children}) => {
  const {_key, level} = value
  return `<li id="${(_key || '').replace('-parent', '')}" data-level="${level}">${children}</li>`
}

const unknownBlockFunc: PortableTextBlockComponent = ({value, children}) =>
  `<p id="${value._key}" data-type="unknown-block-style" data-style="${value.style}">${children}</p>`

export const customSerializers: Record<string, any> = {
  unknownType: ({value}: {value: Record<string, any>}) => `<div class="${value._type}"></div>`,
  types: {},
  block: defaultPortableTextBlockStyles,
  list: defaultLists,
  listItem: defaultListItem,
  unknownBlockStyle: unknownBlockFunc,
}

export const customDeserializers: Record<string, any> = {types: {}}

export const customBlockDeserializers: Array<any> = [
  //handle undeclared styles
  {
    deserialize(
      el: HTMLParagraphElement,
      next: (elements: Node | Node[] | NodeList) => TypedObject | TypedObject[] | undefined
    ): PortableTextTextBlock | TypedObject | undefined {
      if (!el.hasChildNodes()) {
        return undefined
      }

      if (el.getAttribute('data-type') !== 'unknown-block-style') {
        return undefined
      }

      const style = el.getAttribute('data-style') ?? ''
      const block = htmlToBlocks(el.outerHTML, blockContentType)[0]

      return {
        ...block,
        style,
        children: next(el.childNodes),
      }
    },
  },
  //handle list items
  {
    deserialize(
      el: HTMLParagraphElement,
      next: (elements: Node | Node[] | NodeList) => TypedObject | TypedObject[] | undefined
    ): PortableTextTextBlock | TypedObject | undefined {
      if (!el.hasChildNodes()) {
        return undefined
      }

      if (el.tagName.toLowerCase() !== 'li') {
        return undefined
      }

      const tagsToStyle: Record<string, string> = {
        ul: 'bullet',
        ol: 'number',
      }

      const parent = el.parentNode as HTMLUListElement | HTMLOListElement
      if (!parent || !parent.tagName) {
        return undefined
      }

      const listItem = tagsToStyle[parent.tagName.toLowerCase()]
      if (!listItem) {
        return undefined
      }

      const level =
        el.getAttribute('data-level') && parseInt(el.getAttribute('data-level') || '0', 10)
      const _key = el.id
      let block = htmlToBlocks(parent.outerHTML, blockContentType)[0]
      const customStyle = el.children?.[0]?.getAttribute('data-style')

      //check if the object inside is also serialized -- that means it has a style
      //or custom annotation and we should use childNode serialization
      const regex = new RegExp(/<("[^"]*"|'[^']*'|[^'">])*>/)
      if (regex.test(el.innerHTML)) {
        const newBlock = htmlToBlocks(el.innerHTML, blockContentType)[0]
        if (newBlock) {
          block = {
            ...block,
            ...newBlock,
            style: customStyle ?? (newBlock as PortableTextTextBlock).style,
          }

          //next(childNodes) plays poorly with custom styles, issue to be filed.
          if (customStyle) {
            return block as PortableTextTextBlock
          }
        }
      }

      return {
        ...block,
        level,
        _key,
        listItem,
        children: next(el.childNodes),
      }
    },
  },
]
