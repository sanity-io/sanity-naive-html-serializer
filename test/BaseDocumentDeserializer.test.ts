import { readFileSync } from 'fs'

import {
  addedCustomDeserializers,
  getDeserialized,
  toPlainText,
  addedCustomSerializers,
  addedBlockDeserializers,
} from './helpers'
import { Block } from '@sanity/types'
import {
  BaseDocumentDeserializer,
  BaseDocumentSerializer,
  defaultStopTypes,
} from '../src'
import { customBlockDeserializers } from '../src/BaseSerializationConfig'

const documentLevelArticle = require('./__fixtures__/documentLevelArticle')
const inlineDocumentLevelArticle = require('./__fixtures__/inlineDocumentLevelArticle')
const fieldLevelArticle = require('./__fixtures__/fieldLevelArticle')
const annotationAndInlineBlocks = require('./__fixtures__/annotationAndInlineBlocks')
const customStyles = require('./__fixtures__/customStyles')

const schema = require('./__fixtures__/schema')
const inlineSchema = require('./__fixtures__/inlineSchema')

let mockTestKey = 0

//needed to make snapshots happy on internal spans (where we don't track keys)
jest.mock('@sanity/block-tools/lib/util/randomKey.js', () => {
  return jest.fn().mockImplementation(() => {
    return `randomKey-${++mockTestKey}`
  })
})

beforeEach(() => {
  mockTestKey = 0
})

test('Global test of working doc-level functionality and snapshot match', () => {
  const deserialized = getDeserialized(documentLevelArticle, 'document')
  expect(deserialized).toMatchSnapshot()
})

test('Global test of working field-level functionality and snapshot match', () => {
  const deserialized = getDeserialized(fieldLevelArticle, 'field')
  expect(deserialized).toMatchSnapshot()
})

test('Contains id of original document', () => {
  const deserialized = getDeserialized(documentLevelArticle, 'document')
  const id = deserialized._id
  expect(id).toEqual(documentLevelArticle._id)
})

test('Contains rev of original document', () => {
  const deserialized = getDeserialized(documentLevelArticle, 'document')
  const rev = deserialized._rev
  expect(rev).toEqual(documentLevelArticle._rev)
})

test('Contains type of original document', () => {
  const deserialized = getDeserialized(documentLevelArticle, 'document')
  const type = deserialized._type
  expect(type).toEqual(documentLevelArticle._type)
})

/*
 * DOCUMENT LEVEL
 */

/*
 * Top-level plain text
 */

test('String and text types get deserialized correctly at top-level -- document level', () => {
  const deserialized = getDeserialized(documentLevelArticle, 'document')
  expect(deserialized.title).toEqual(documentLevelArticle.title)
  expect(deserialized.snippet).toEqual(documentLevelArticle.snippet)
})

/*
 * Presence and accuracy of fields in "vanilla" deserialization -- objects
 */
test('Nested object contains accurate values -- document level', () => {
  const deserialized = getDeserialized(documentLevelArticle, 'document')
  const origTitle = documentLevelArticle.config.title
  const deserializedTitle = deserialized.config.title
  expect(origTitle).toEqual(deserializedTitle)

  const origBlockText = documentLevelArticle.config.nestedArrayField
  const deserializedBlockText = deserialized.config.nestedArrayField

  const origKeys = origBlockText.map((block: Block) => block._key)
  const deserializedKeys = deserializedBlockText.map(
    (block: Block) => block._key
  )

  expect(deserializedKeys.sort()).toEqual(origKeys.sort())
  expect(toPlainText(deserializedBlockText)).toEqual(toPlainText(origBlockText))
})

test('Nested object in an object contains accurate values -- document level', () => {
  const deserialized = getDeserialized(documentLevelArticle, 'document')
  const origNestedObject = documentLevelArticle.config.objectAsField
  const deserializedNestedObject = deserialized.config.objectAsField

  expect(origNestedObject.title).toEqual(deserializedNestedObject.title)

  const origKeys = origNestedObject.content.map((block: Block) => block._key)
  const deserializedKeys = deserializedNestedObject.content.map(
    (block: Block) => block._key
  )

  expect(origKeys.sort()).toEqual(deserializedKeys.sort())
  expect(toPlainText(deserializedNestedObject.content)).toEqual(
    toPlainText(origNestedObject.content)
  )
})

/*
 * Presence and accuracy of fields in vanilla deserialization -- arrays
 */

test('Array contains all serializable blocks with keys, in order -- document level', () => {
  const deserialized = getDeserialized(documentLevelArticle, 'document')
  const origKeys = documentLevelArticle.content.map(
    (block: Block) => block._key
  )
  const deserializedKeys = deserialized.content.map(
    (block: Block) => block._key
  )
  expect(deserializedKeys.sort()).toEqual(origKeys.sort())
})

test('Array contains top-level block text -- document level', () => {
  const deserialized = getDeserialized(documentLevelArticle, 'document')
  expect(toPlainText(deserialized.content)).toEqual(
    toPlainText(documentLevelArticle.content)
  )
})

test('Object in array contains accurate values in nested object -- document level', () => {
  const deserialized = getDeserialized(documentLevelArticle, 'document')
  const origTitle = documentLevelArticle.content.find(
    (block: Record<string, any>) => block._type === 'objectField'
  ).objectAsField.title
  const deserializedTitle = deserialized.content.find(
    (block: Record<string, any>) => block._type === 'objectField'
  ).objectAsField.title
  expect(deserializedTitle).toEqual(origTitle)

  const origBlockText = toPlainText(
    documentLevelArticle.content.find(
      (block: Record<string, any>) => block._type === 'objectField'
    ).objectAsField.content
  ).trim()
  const deserializedBlockText = toPlainText(
    documentLevelArticle.content.find(
      (block: Record<string, any>) => block._type === 'objectField'
    ).objectAsField.content
  ).trim()
  expect(deserializedBlockText).toEqual(origBlockText)
})

/*
 * FIELD LEVEL
 */

/*
 * Top-level plain text
 */

test('String and text types get deserialized correctly at top-level -- field level', () => {
  const deserialized = getDeserialized(fieldLevelArticle, 'field')
  expect(deserialized.title.en).toEqual(fieldLevelArticle.title.en)
  expect(deserialized.snippet.en).toEqual(fieldLevelArticle.snippet.en)
})

/*
 * Presence and accuracy of fields in "vanilla" deserialization -- objects
 */
test('Nested object contains accurate values -- field level', () => {
  const deserialized = getDeserialized(fieldLevelArticle, 'field')
  const origTitle = fieldLevelArticle.config.en.title
  const deserializedTitle = deserialized.config.en.title
  expect(origTitle).toEqual(deserializedTitle)

  const origBlockText = fieldLevelArticle.config.en.nestedArrayField
  const deserializedBlockText = deserialized.config.en.nestedArrayField

  const origKeys = origBlockText.map((block: Block) => block._key)
  const deserializedKeys = deserializedBlockText.map(
    (block: Block) => block._key
  )

  expect(deserializedKeys.sort()).toEqual(origKeys.sort())
  expect(toPlainText(deserializedBlockText)).toEqual(toPlainText(origBlockText))
})

test('Nested object in an object contains accurate values -- field level', () => {
  const deserialized = getDeserialized(fieldLevelArticle, 'field')
  const origNestedObject = fieldLevelArticle.config.en.objectAsField
  const deserializedNestedObject = deserialized.config.en.objectAsField

  expect(origNestedObject.title).toEqual(deserializedNestedObject.title)

  const origKeys = origNestedObject.content.map((block: Block) => block._key)
  const deserializedKeys = deserializedNestedObject.content.map(
    (block: Block) => block._key
  )

  expect(origKeys.sort()).toEqual(deserializedKeys.sort())
  expect(toPlainText(deserializedNestedObject.content)).toEqual(
    toPlainText(origNestedObject.content)
  )
})

/*
 * Presence and accuracy of fields in vanilla deserialization -- arrays
 */

test('Array contains all serializable blocks with keys, in order -- field level', () => {
  const deserialized = getDeserialized(fieldLevelArticle, 'field')
  const origKeys = fieldLevelArticle.content.en.map(
    (block: Block) => block._key
  )
  const deserializedKeys = deserialized.content.en.map(
    (block: Block) => block._key
  )
  expect(deserializedKeys.sort()).toEqual(origKeys.sort())
})

test('Array contains top-level block text -- field level', () => {
  const deserialized = getDeserialized(fieldLevelArticle, 'field')
  expect(toPlainText(deserialized.content.en)).toEqual(
    toPlainText(fieldLevelArticle.content.en)
  )
})

test('Object in array contains accurate values in nested object -- document level', () => {
  const deserialized = getDeserialized(fieldLevelArticle, 'field')
  const origTitle = fieldLevelArticle.content.en.find(
    (block: Record<string, any>) => block._type === 'objectField'
  ).objectAsField.title
  const deserializedTitle = deserialized.content.en.find(
    (block: Record<string, any>) => block._type === 'objectField'
  ).objectAsField.title
  expect(deserializedTitle).toEqual(origTitle)

  const origBlockText = toPlainText(
    fieldLevelArticle.content.en.find(
      (block: Record<string, any>) => block._type === 'objectField'
    ).objectAsField.content
  ).trim()
  const deserializedBlockText = toPlainText(
    fieldLevelArticle.content.en.find(
      (block: Record<string, any>) => block._type === 'objectField'
    ).objectAsField.content
  ).trim()
  expect(deserializedBlockText).toEqual(origBlockText)
})

/*
 * CUSTOM SETTINGS
 */

test('Custom deserialization should manifest at all levels', () => {
  const serialized = BaseDocumentSerializer(schema).serializeDocument(
    documentLevelArticle,
    'document',
    'en',
    defaultStopTypes,
    addedCustomSerializers
  )

  const deserialized = BaseDocumentDeserializer.deserializeDocument(
    serialized.content,
    addedCustomDeserializers,
    customBlockDeserializers
  )
  expect(deserialized.config.title).toEqual(documentLevelArticle.config.title)
  expect(deserialized.config._type).toEqual(documentLevelArticle.config._type)

  const origArrayObj = documentLevelArticle.content.find(
    (b: Record<string, any>) => b._type === 'objectField'
  )
  const deserializedArrayObj = deserialized.content.find(
    (b: Record<string, any>) => b._type === 'objectField'
  )

  expect(deserializedArrayObj.title).toEqual(origArrayObj.title)
  expect(deserializedArrayObj._key).toEqual(origArrayObj._key)
})

//test -- unhandled annotations and inlines don't break when they get deserialized back
test('Handled inline objects should be accurately deserialized', () => {
  const inlineDocument = {
    ...documentLevelArticle,
    ...annotationAndInlineBlocks,
  }

  const serialized = BaseDocumentSerializer(schema).serializeDocument(
    inlineDocument,
    'document',
    'en',
    defaultStopTypes,
    addedCustomSerializers
  )

  const deserialized = BaseDocumentDeserializer.deserializeDocument(
    serialized.content,
    addedCustomDeserializers,
    addedBlockDeserializers
  )

  let origInlineObj: Record<string, any> | null = null
  let deserializedInlineObj: Record<string, any> | null = null

  inlineDocument.content.forEach((block: Record<string, any>) => {
    if (block.children) {
      block.children.forEach((span: Record<string, any>) => {
        if (span._type === 'childObjectField') {
          origInlineObj = span
        }
      })
    }
  })

  deserialized.content.forEach((block: Record<string, any>) => {
    if (block.children) {
      block.children.forEach((span: Record<string, any>) => {
        if (span._type === 'childObjectField') {
          deserializedInlineObj = span
        }
      })
    }
  })

  expect(deserializedInlineObj!.title).toEqual(origInlineObj!.title)
  expect(deserializedInlineObj!._type).toEqual(origInlineObj!._type)
})

test('Handled annotations should be accurately deserialized', () => {
  const inlineDocument = {
    ...documentLevelArticle,
    ...annotationAndInlineBlocks,
  }

  const serialized = BaseDocumentSerializer(schema).serializeDocument(
    inlineDocument,
    'document',
    'en',
    defaultStopTypes,
    addedCustomSerializers
  )

  const deserialized = BaseDocumentDeserializer.deserializeDocument(
    serialized.content,
    addedCustomDeserializers,
    addedBlockDeserializers
  )

  let origAnnotation: Record<string, any> | null = null
  let deserializedAnnotation: Record<string, any> | null = null

  inlineDocument.content.forEach((block: Block) => {
    if (block.children) {
      block.children.forEach((span: Record<string, any>) => {
        if (span.marks && span.marks.length) {
          origAnnotation = span
        }
      })
    }
  })

  deserialized.content.forEach((block: Block) => {
    if (block.children) {
      block.children.forEach((span: Record<string, any>) => {
        if (span.marks && span.marks.length) {
          deserializedAnnotation = span
        }
      })
    }
  })

  expect(deserializedAnnotation!.text).toEqual(origAnnotation!.text)
})

/*
 * STYLE TAGS
 */
test('Deserialized content should preserve style tags', () => {
  const deserialized = getDeserialized(documentLevelArticle, 'document')
  const origH1 = documentLevelArticle.content.find(
    (block: Block) => block.style === 'h1'
  )
  const deserializedH1 = deserialized.content.find(
    (block: Block) => block.style === 'h1'
  )
  const origH2 = documentLevelArticle.content.find(
    (block: Block) => block.style === 'h2'
  )
  const deserializedH2 = deserialized.content.find(
    (block: Block) => block.style === 'h2'
  )
  expect(deserializedH1).toBeDefined()
  expect(deserializedH2).toBeDefined()
  expect(deserializedH1._key).toEqual(origH1._key)
  expect(deserializedH2._key).toEqual(origH2._key)
  expect(deserializedH1.children[0].text).toEqual(origH1.children[0].text)
  expect(deserializedH2.children[0].text).toEqual(origH2.children[0].text)
})

test('Content with custom styles deserializes correctly and maintains style', () => {
  //unhandled style will throw a warn -- ignore it in this case
  jest.spyOn(console, 'warn').mockImplementation(() => {})

  const customStyledDocument = {
    ...documentLevelArticle,
    ...customStyles,
  }

  const serialized = BaseDocumentSerializer(schema).serializeDocument(
    customStyledDocument,
    'document'
  )

  const deserialized = BaseDocumentDeserializer.deserializeDocument(
    serialized.content
  )

  expect(deserialized.content[0].children[0].text).toEqual(
    customStyledDocument.content[0].children[0].text
  )

  expect(deserialized.content[0].style).toEqual(
    customStyledDocument.content[0].style
  )
})

/*
 * MESSY INPUT
 */
test('&nbsp; whitespace should not be escaped', () => {
  //unhandled field will throw a warn -- ignore it in this case
  jest.spyOn(console, 'debug').mockImplementation(() => {})

  const content = readFileSync('test/__fixtures__/messy-html.html', {
    encoding: 'utf-8',
  })
  const result = BaseDocumentDeserializer.deserializeDocument(content)
  expect(result.title).toEqual('Här är artikel titeln')
  expect(result.content[1].nestedArrayField[0].title).toEqual(
    'Det här är en dragspels titeln'
  )
})

/*
 * V2 functionality -- be able to operate without a strict schema
 */
test('Content with anonymous inline objects deserializes all fields, at any depth', () => {
  //unhandled field will throw a warn -- ignore it in this case
  jest.spyOn(console, 'debug').mockImplementation(() => {})

  const serialized = BaseDocumentSerializer(inlineSchema).serializeDocument(
    inlineDocumentLevelArticle,
    'document'
  )

  const deserialized = BaseDocumentDeserializer.deserializeDocument(
    serialized.content
  )
  //object in field
  expect(deserialized.tabs.config.title).toEqual(
    inlineDocumentLevelArticle.tabs.config.title
  )

  //array in object in object
  expect(
    deserialized.tabs.config.objectAsField.content[0].children[0].text
  ).toEqual(
    inlineDocumentLevelArticle.tabs.config.objectAsField.content[0].children[0]
      .text
  )

  //arrays
  expect(deserialized.tabs.content).toBeInstanceOf(Array)
  expect(deserialized.tabs.content.map((block: any) => block._key)).toEqual(
    inlineDocumentLevelArticle.tabs.content.map((block: any) => block._key)
  )

  //object in array
  const origObj = inlineDocumentLevelArticle.tabs.content.find(
    (block: any) => block._type === 'objectField'
  )
  const deserializedObj = deserialized.tabs.content.find(
    (block: any) => block._type === 'objectField'
  )

  expect(deserializedObj.title).toEqual(origObj.title)
  expect(deserializedObj.objectAsField.content[0].children[0].text).toEqual(
    origObj.objectAsField.content[0].children[0].text
  )

  //anonymous object in array
  const origArray = inlineDocumentLevelArticle.tabs.arrayWithAnonymousObjects
  const deserializedArray = deserialized.tabs.arrayWithAnonymousObjects
  expect(deserializedArray.length).toEqual(origArray.length)
  expect(deserializedArray[0]._key).toEqual(origArray[0]._key)
  expect(Object.keys(deserializedArray[0])).not.toContain('span')
})
