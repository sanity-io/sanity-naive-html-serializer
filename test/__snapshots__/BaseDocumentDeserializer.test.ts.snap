// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Global test of working doc-level functionality and snapshot match 1`] = `
Object {
  "_id": "drafts.d8ffc675-ce86-4f60-9ac8-da164cde3b0a",
  "_rev": "fxcnqj-uew-gr7-v6r-z965tfm69",
  "_type": "documentLevelArticle",
  "config": Object {
    "_type": "objectField",
    "nestedArrayField": Array [
      Object {
        "_key": "4a58adc7c507",
        "_type": "block",
        "children": Array [
          Object {
            "_key": "randomKey-8",
            "_type": "span",
            "marks": Array [],
            "text": "This is block text 1 level deep",
          },
        ],
        "markDefs": Array [],
        "style": "normal",
      },
    ],
    "objectAsField": Object {
      "_type": "childObjectField",
      "content": Array [
        Object {
          "_key": "1aa228ac848b",
          "_type": "block",
          "children": Array [
            Object {
              "_key": "randomKey-7",
              "_type": "span",
              "marks": Array [],
              "text": "This is a block text 2 levels deep",
            },
          ],
          "markDefs": Array [],
          "style": "normal",
        },
      ],
      "title": "This is one level deeper",
    },
    "title": "This is an object nested in a document",
  },
  "content": Array [
    Object {
      "_key": "a4d177e92666",
      "_type": "block",
      "children": Array [
        Object {
          "_key": "randomKey-9",
          "_type": "span",
          "marks": Array [],
          "text": "This is block text at the top level.",
        },
      ],
      "markDefs": Array [],
      "style": "normal",
    },
    Object {
      "_key": "c0313627775e",
      "_type": "objectField",
      "objectAsField": Object {
        "_type": "childObjectField",
        "content": Array [
          Object {
            "_key": "4f24f6fbfae7",
            "_type": "block",
            "children": Array [
              Object {
                "_key": "randomKey-12",
                "_type": "span",
                "marks": Array [],
                "text": "This is block text in a nested object in an object in top-level block text.",
              },
            ],
            "markDefs": Array [],
            "style": "normal",
          },
        ],
        "title": "This is a nested object in an object in top-level block text.",
      },
      "title": "This is an object in top-level block text.",
    },
    Object {
      "_key": "9e2ab13c6d63",
      "_type": "block",
      "children": Array [
        Object {
          "_key": "randomKey-13",
          "_type": "span",
          "marks": Array [],
          "text": "This is h1 text",
        },
      ],
      "markDefs": Array [],
      "style": "h1",
    },
    Object {
      "_key": "297142fbaf21",
      "_type": "block",
      "children": Array [
        Object {
          "_key": "randomKey-14",
          "_type": "span",
          "marks": Array [],
          "text": "This is h2 text",
        },
      ],
      "markDefs": Array [],
      "style": "h2",
    },
    Object {
      "_key": "76e648fc3845",
      "_type": "block",
      "children": Array [
        Object {
          "_key": "randomKey-16",
          "_type": "span",
          "marks": Array [],
          "text": "Bullet 1",
        },
      ],
      "level": 1,
      "listItem": "bullet",
      "markDefs": Array [],
      "style": "normal",
    },
    Object {
      "_key": "d090cd8b27d2",
      "_type": "block",
      "children": Array [
        Object {
          "_key": "randomKey-18",
          "_type": "span",
          "marks": Array [],
          "text": "nested bullet a",
        },
      ],
      "level": 2,
      "listItem": "bullet",
      "markDefs": Array [],
      "style": "normal",
    },
    Object {
      "_key": "1cdffa5d50f5",
      "_type": "block",
      "children": Array [
        Object {
          "_key": "randomKey-21",
          "_type": "span",
          "marks": Array [],
          "text": "Bullet 2",
        },
      ],
      "level": 1,
      "listItem": "bullet",
      "markDefs": Array [],
      "style": "h2",
    },
  ],
  "snippet": "This is text in my text field.",
  "tags": Array [
    "tag 1",
    "tag 2",
    "tag 3",
  ],
  "title": "My Document-Level Article",
}
`;

exports[`Global test of working field-level functionality and snapshot match 1`] = `
Object {
  "_id": "drafts.2947533e-1ea5-4116-955b-339608d3445d",
  "_rev": "l58oha-n06-1s4-f6i-74g6cuakl",
  "_type": "fieldLevelArticle",
  "config": Object {
    "en": Object {
      "_type": "objectField",
      "nestedArrayField": Array [
        Object {
          "_key": "f49b4d7e3e51",
          "_type": "block",
          "children": Array [
            Object {
              "_key": "randomKey-8",
              "_type": "span",
              "marks": Array [],
              "text": "This is block text 1 level deep",
            },
          ],
          "markDefs": Array [],
          "style": "normal",
        },
      ],
      "objectAsField": Object {
        "_type": "childObjectField",
        "content": Array [
          Object {
            "_key": "869168fe3a2a",
            "_type": "block",
            "children": Array [
              Object {
                "_key": "randomKey-7",
                "_type": "span",
                "marks": Array [],
                "text": "This is block text 2 levels deep",
              },
            ],
            "markDefs": Array [],
            "style": "normal",
          },
        ],
        "title": "This is one level deeper",
      },
      "title": "This is an object nested in a document",
    },
  },
  "content": Object {
    "en": Array [
      Object {
        "_key": "e2a39d768ff8",
        "_type": "block",
        "children": Array [
          Object {
            "_key": "randomKey-9",
            "_type": "span",
            "marks": Array [],
            "text": "This is block text at the top level.",
          },
        ],
        "markDefs": Array [],
        "style": "normal",
      },
      Object {
        "_key": "271b0c6ee984",
        "_type": "objectField",
        "objectAsField": Object {
          "_type": "childObjectField",
          "content": Array [
            Object {
              "_key": "4f24f6fbfae7",
              "_type": "block",
              "children": Array [
                Object {
                  "_key": "randomKey-12",
                  "_type": "span",
                  "marks": Array [],
                  "text": "This is block text in a nested object in an object in top-level block text.",
                },
              ],
              "markDefs": Array [],
              "style": "normal",
            },
          ],
          "title": "This is a nested object in an object in top-level block text.",
        },
        "title": "This is an object in top-level block text.",
      },
    ],
  },
  "snippet": Object {
    "en": "This is text in my text field",
  },
  "tags": Object {
    "en": Array [
      "tag 1",
      "tag 2",
      "tag 3",
    ],
  },
  "title": Object {
    "en": "My Field-Level Article",
  },
}
`;
